### PR TITLE
Adding tutorial for dinitz algorithm

### PR DESCRIPTION
**Pull request by harshal-dupare**

### CODE REVIEW

File: content/algorithms/flow/dinitz_alg.md
Comment (rossbar) [inline content/algorithms/flow/dinitz_alg.md:40]:
````suggestion
```{code-cell} ipython3
---
tags: [hide-input]
---
# code to generate the first image goes here!
```
````

**Review by rossbar — COMMENTED** [review]
Generally this looks really good - the notebook flows very nicely and the examples are very illustrative!

As discussed, one of the major difficulties is figuring out how to have the figures produced via code rather loaded from file. AIUI the code to generate all of the figures currently lives in a separate repo. IMO it would definitely be better to have that code incorporated in the notebook example itself both for reproducibility and pedagogical purposes. I definitely understand the main motivation for not doing so --- the code is very long and mostly deals with visualization, so it distracts from the main point of the tutorial.

Fortunately I think there's a good solution here - we can use the [hiding cells feature of myst-nb](https://myst-nb.readthedocs.io/en/latest/use/hiding.html#hiding-code-cells). This has the advantage that the code is still incorporated in the notebook and the images are produced during notebook execution which guarantees everything stays synced up; but the code itself is "hidden" (but still accessible by clicking an expansion button) in the html output by default so that the main narrative isn't disturbed. The code cells themselves are still present in the notebook so they will be visible to anyone using the NBs interactively (which I'd argue is what we want) - but it keeps the html version of the tutorial nice and clean.

It'd be great if we could give this a try. In order not to get completely bogged down, it might make sense to try this out with a single image first. See for example the inline comment I left. LMK what you think!

File: content/algorithms/flow/dinitz_alg.md
Comment (rossbar) [inline content/algorithms/flow/dinitz_alg.md]:
```suggestion
Note that dinitz works in $O(n^2m)$ strongly polynomial time, that is to say it 
doesn't depend on the value of flow. It is noteworthy that its performance of biparted 
```
The wording here was a bit off... I *think* the suggestion captures what you were trying to say, but I could be wrong. Please correct me as you see fit!

File: content/algorithms/flow/dinitz_alg.md
Comment (rossbar) [inline content/algorithms/flow/dinitz_alg.md]:
```suggestion
Lets consider the example of shipping packages from warehouse to customers.
The packages must be routed through some 
intermediate shipping points, and we can only ship so many packages per day through 
any given point.
```

File: content/algorithms/flow/dinitz_alg.md
Comment (rossbar) [inline content/algorithms/flow/dinitz_alg.md]:
```suggestion
Given these constraints, the shipping company would like to assign intermediate shipping points to customers so that the maximum number of packages are shipped each day.
```

File: content/algorithms/flow/dinitz_alg.md
Comment (rossbar) [inline content/algorithms/flow/dinitz_alg.md]:
```suggestion
The number below each intermediate shipping point is the maximum number of shipments that 
can be processed per day, and the edges represent which customers can be reached from a
given shipping point.
```

File: content/algorithms/flow/dinitz_alg.md
Comment (rossbar) [inline content/algorithms/flow/dinitz_alg.md]:
```suggestion
Let's label the original warehouse $W$, the intermediate shipping points as 
$lw1, lw2, lw3$, and customers as $c1,c2...c20$.
```

**Review by rossbar — COMMENTED** [review]
I had a chance to look at the notebook in a little more detail - this is great @harshal-dupare !

I mostly focused on the text in this review, so most of the comments are related to wording/minor organization suggestions. Feel free to address them as you see fit - no need to accept the suggestions as-is.

Other than that, this is in a good state, though I'd like to hold off on merging until we've had time to try to generate the figures via execution. The reason for this is: this tutorial adds ~20+ MB of files (most notably the gif) to the repo, which increases the size quite a bit. Now that the code is all in the notebook (though not yet necessarily being run) I'd like to try to get the images into a fully reproducible state to see if we can manage the size. It may end up that this is too much effort, but I'd like to give it a shot before merging!

**Comment by rossbar** [comment]
To summarize the situation: the tutorial is excellent, but the main blocker to merging is that merging the PR would be adding 20ish MB to the commit history because all of the images (particularly the gif) are pre-computed elsewhere and linked to in the notebook, rather than generated during execution. Adding the code for generating the gif in a non-executed cell was an improvement, but didn't help on the commit-size front since the images were all loaded from file

I spent some time to refactor the notebook so that now the images are generated during execution. It was a fair amount of work, but in the end I think it's necessary to help move this across the finish line! Note that in the end I didn't end up replacing the gif - all of the pieces are in place to do so, but it would be a significant task that I thought could be left for a followup.

**Comment by MridulS** [comment]
Thanks for you work on this @harshal-dupare @rossbar! This is great.
I went through this once, and it looks good to go, we can iterate on this further if required in a new PR :)

**Review by MridulS — APPROVED** [review]

**Pull request merged**
